File: src/msu_hub_bot/web/feedback.py
"""Reviewer-only HTTP projections of privately stored feedback reports."""

from __future__ import annotations

import re
from typing import TYPE_CHECKING
from uuid import UUID

from aiohttp import web
from pydantic import BaseModel, ConfigDict, Field

from msu_hub_bot.feedback.models import (
    FeedbackAccessDenied,
    FeedbackKind,
    FeedbackMessage,
    FeedbackReport,
    FeedbackReview,
    FeedbackReviewStatus,
    SelectedFeedbackContext,
)
from msu_hub_bot.feedback.service import FeedbackService
from msu_hub_bot.storage.features import Record

if TYPE_CHECKING:
    from .server import WebServer


class ReviewInput(BaseModel):
    model_config = ConfigDict(extra="forbid", hide_input_in_errors=True)

    etag: UUID
    status: FeedbackReviewStatus
    note: str = Field(max_length=2000)


def _message(value: FeedbackMessage | None) -> dict[str, object] | None:
    if value is None:
        return None
    return value.model_dump(
        mode="json",
        include={
            "chat_id",
            "message_id",
            "sent_at",
            "author_name",
            "text",
            "thread_id",
            "author_id",
            "author_kind",
            "media_kind",
            "truncated",
        },
    )


def _context(value: SelectedFeedbackContext) -> dict[str, object]:
    return {
        "origin": value.origin.model_dump(mode="json", include={"chat_id", "thread_id", "label"}) if value.origin else None,
        "reply": _message(value.reply),
        "recent_messages": [_message(message) for message in value.recent_messages],
        "diagnostics": [
            item.model_dump(mode="json", include={"at", "handler", "outcome", "command", "message_id", "trace_id", "reason", "release"})
            for item in value.diagnostics
        ],
        "recent_available": value.recent_available,
        "reply_available": value.reply_available,
        "diagnostics_since": value.diagnostics_since.isoformat() if value.diagnostics_since else None,
    }


def _review(record: Record[FeedbackReview], *, detail: bool) -> dict[str, object]:
    fields = {
        "report_id",
        "author_id",
        "author_name",
        "summary",
        "kind",
        "created_at",
        "submitted_at",
        "status",
        "reviewer_id",
        "reviewed_at",
    }
    if detail:
        fields.add("note")
    return {**record.value.model_dump(mode="json", include=fields), "etag": record.etag}


def _report(record: Record[FeedbackReport]) -> dict[str, object]:
    value = record.value
    return {
        **value.model_dump(
            mode="json",
            include={"report_id", "author_id", "author_name", "created_at", "submitted_at", "kind", "description", "rendered_text"},
        ),
        "context": _context(value.context),
        "delivery": value.model_dump(
            mode="json", include={"status", "attempts", "sending_at", "sent_at", "delivered_message_id", "failure"}
        ),
    }


class FeedbackAPI:
    def __init__(self, server: WebServer, service: FeedbackService | None) -> None:
        self.server, self.service = server, service

    def register(self, app: web.Application) -> None:
        app.router.add_get("/api/feedback", self.list)
        app.router.add_get("/api/feedback/{report_id}", self.detail)
        app.router.add_post("/api/feedback/{report_id}/review", self.review)

    def allowed(self, user_id: int) -> bool:
        return self.service is not None and self.service.is_reviewer(user_id)

    def _authorize(self, request: web.Request) -> tuple[FeedbackService, int]:
        from .server import USER

        user_id = request[USER].id
        if self.service is None or not self.service.is_reviewer(user_id):
            raise FeedbackAccessDenied()
        return self.service, user_id

    @staticmethod
    def _key(request: web.Request) -> str:
        key = request.match_info["report_id"]
        if re.fullmatch(r"[a-f0-9]{16}", key) is None:
            raise ValueError("Invalid feedback identifier")
        return key

    async def list(self, request: web.Request) -> web.Response:
        service, reviewer_id = self._authorize(request)
        query = request.query
        if set(query) - {"status", "kind", "after", "limit"} or len(query) != len(set(query)):
            raise ValueError("Invalid feedback query")
        status: FeedbackReviewStatus | None = None
        if "status" in query:
            match query["status"]:
                case "new" | "in_progress" | "done" | "dismissed" as selected:
                    status = selected
                case _:
                    raise ValueError("Invalid feedback status")
        kind: FeedbackKind | None = None
        if "kind" in query:
            match query["kind"]:
                case "bug" | "idea" | "other" as selected_kind:
                    kind = selected_kind
                case _:
                    raise ValueError("Invalid feedback kind")
        after = query.get("after")
        if after is not None and re.fullmatch(r"[0-9]{16}:[a-f0-9]{16}", after) is None:
            raise ValueError("Invalid feedback cursor")
        raw_limit = query.get("limit", "20")
        if re.fullmatch(r"[1-9][0-9]?", raw_limit) is None or int(raw_limit) > 50:
            raise ValueError("Invalid feedback page size")
        limit = int(raw_limit)
        records = await service.review_list(reviewer_id, status=status, kind=kind, after=after, limit=limit)
        return web.json_response(
            {
                "items": [_review(record, detail=False) for record in records],
                "next_cursor": records[-1].key if len(records) == limit else None,
            }
        )

    async def detail(self, request: web.Request) -> web.Response:
        service, reviewer_id = self._authorize(request)
        if request.query:
            raise ValueError("Feedback detail does not accept query parameters")
        report, review = await service.review_get(reviewer_id, self._key(request))
        return web.json_response({"report": _report(report), "review": _review(review, detail=True)})

    async def review(self, request: web.Request) -> web.Response:
        service, reviewer_id = self._authorize(request)
        if request.query:
            raise ValueError("Feedback review does not accept query parameters")
        key = self._key(request)
        body = await self.server._body(request, ReviewInput)
        record = await service.review_update(reviewer_id, key, expected_etag=str(body.etag), status=body.status, note=body.note)
        return web.json_response(_review(record, detail=True))

File: web/src/platform/feedback.ts
import { ApiClient, ApiError } from "./api";
import { array, boolean, integer, object, stamp, string } from "./decode";
import type {
  DeliveryStatus,
  FeedbackContext,
  FeedbackDetail,
  FeedbackFilter,
  FeedbackKind,
  FeedbackMessage,
  FeedbackPage,
  FeedbackReview,
  FeedbackSummary,
  ReviewStatus,
} from "./feedbackTypes";

function choice<T extends string>(value: unknown, choices: readonly T[]): T {
  const result = string(value);
  if (!choices.includes(result as T))
    throw new ApiError(
      "protocol",
      "Не удалось прочитать состояние отзыва.",
      503,
    );
  return result as T;
}
function nullable<T>(value: unknown, decode: (value: unknown) => T): T | null {
  return value === null ? null : decode(value);
}
function kind(value: unknown): FeedbackKind {
  return choice(value, ["bug", "idea", "other"]);
}
function summary(value: unknown): FeedbackSummary {
  const row = object(value);
  return {
    report_id: string(row.report_id),
    author_id: integer(row.author_id),
    author_name: string(row.author_name),
    summary: string(row.summary),
    kind: kind(row.kind),
    created_at: stamp(row.created_at),
    submitted_at: stamp(row.submitted_at),
    status: choice<ReviewStatus>(row.status, [
      "new",
      "in_progress",
      "done",
      "dismissed",
    ]),
    reviewer_id: nullable(row.reviewer_id, integer),
    reviewed_at: nullable(row.reviewed_at, stamp),
    etag: string(row.etag),
  };
}
function review(value: unknown): FeedbackReview {
  return { ...summary(value), note: string(object(value).note) };
}
function message(value: unknown): FeedbackMessage {
  const row = object(value);
  return {
    chat_id: integer(row.chat_id),
    message_id: integer(row.message_id),
    thread_id: nullable(row.thread_id, integer),
    author_id: nullable(row.author_id, integer),
    author_name: string(row.author_name),
    sent_at: stamp(row.sent_at),
    text: string(row.text),
    media_kind: nullable(row.media_kind, string),
    truncated: boolean(row.truncated),
  };
}
function context(value: unknown): FeedbackContext {
  const row = object(value);
  return {
    origin: nullable(row.origin, (value) => {
      const origin = object(value);
      return {
        chat_id: integer(origin.chat_id),
        thread_id: nullable(origin.thread_id, integer),
        label: string(origin.label),
      };
    }),
    reply: nullable(row.reply, message),
    recent_messages: array(row.recent_messages, message),
    recent_available: boolean(row.recent_available),
    reply_available: boolean(row.reply_available),
    diagnostics_since: nullable(row.diagnostics_since, stamp),
    diagnostics: array(row.diagnostics, (value) => {
      const entry = object(value);
      return {
        at: stamp(entry.at),
        handler: string(entry.handler),
        outcome: string(entry.outcome),
        command: nullable(entry.command, string),
        message_id: nullable(entry.message_id, integer),
        trace_id: nullable(entry.trace_id, string),
        reason: nullable(entry.reason, string),
        release: nullable(entry.release, string),
      };
    }),
  };
}
function detail(value: unknown): FeedbackDetail {
  const row = object(value),
    report = object(row.report),
    delivery = object(report.delivery);
  return {
    report: {
      report_id: string(report.report_id),
      author_id: integer(report.author_id),
      author_name: string(report.author_name),
      created_at: stamp(report.created_at),
      submitted_at: stamp(report.submitted_at),
      kind: kind(report.kind),
      description: string(report.description),
      rendered_text: string(report.rendered_text),
      context: context(report.context),
      delivery: {
        status: choice<DeliveryStatus>(delivery.status, [
          "queued",
          "sending",
          "sent",
          "uncertain",
          "failed",
        ]),
        attempts: integer(delivery.attempts),
        sending_at: nullable(delivery.sending_at, stamp),
        sent_at: nullable(delivery.sent_at, stamp),
        delivered_message_id: nullable(delivery.delivered_message_id, integer),
        failure: nullable(delivery.failure, string),
      },
    },
    review: review(row.review),
  };
}
export class FeedbackApi extends ApiClient {
  feedbackList(
    filters: FeedbackFilter,
    after?: string,
    signal?: AbortSignal,
  ): Promise<FeedbackPage> {
    const query = new URLSearchParams({ limit: "30" });
    if (filters.status) query.set("status", filters.status);
    if (filters.kind) query.set("kind", filters.kind);
    if (after) query.set("after", after);
    return this.request(
      `/api/feedback?${query}`,
      (value) => {
        const row = object(value);
        return {
          items: array(row.items, summary),
          next_cursor: nullable(row.next_cursor, string),
        };
      },
      undefined,
      signal,
    );
  }
  feedbackDetail(id: string, signal?: AbortSignal): Promise<FeedbackDetail> {
    return this.request(
      `/api/feedback/${encodeURIComponent(id)}`,
      detail,
      undefined,
      signal,
    );
  }
  saveFeedbackReview(
    id: string,
    body: { etag: string; status: ReviewStatus; note: string },
  ): Promise<FeedbackReview> {
    return this.request(
      `/api/feedback/${encodeURIComponent(id)}/review`,
      review,
      body,
    );
  }
}
